src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java

```diff
@@ -0,0 +1,440 @@
+                preparedAssignments.add(namedSelect);
+                refSources.put(name, new SelectReferenceSource(prepared));
+            }
+
+            if (dataReferences != null)
+                for (RowDataReference.Raw reference : dataReferences)
+                    reference.resolveReference(refSources);
+
+            NamedSelect returningSelect = null;
+            if (select != null)
+            {
+                SelectStatement prepared = select.prepare(bindVariables);
+                // TODO: Accord saves the result of this read, so limit to a single row until that is no longer true.
+                checkAtMostOneRowSpecified(prepared, INCOMPLETE_PRIMARY_KEY_SELECT_MESSAGE, LazyToString.lazy(() -> prepared.asCQL(QueryOptions.DEFAULT, state)));
+                returningSelect = new NamedSelect(TxnDataName.returning(), prepared);
+            }
+
+            List<RowDataReference> returningReferences = null;
+
+            if (returning != null)
+            {
+                // TODO: Eliminate/modify this check if we allow full tuple selections.
+                returningReferences = returning.stream().peek(raw -> checkTrue(raw.column() != null, SELECT_REFS_NEED_COLUMN_MESSAGE))
+                                                        .map(RowDataReference.Raw::prepareAsReceiver)
+                                                        .collect(Collectors.toList());
+            }
+
+            List<ModificationStatement> preparedUpdates = new ArrayList<>(updates.size());
+            
+            // check for any read-before-write updates
+            for (int i = 0; i < updates.size(); i++)
+            {
+                ModificationStatement.Parsed parsed = updates.get(i);
+
+                ModificationStatement prepared = parsed.prepare(bindVariables);
+                checkFalse(prepared.hasConditions(), NO_CONDITIONS_IN_UPDATES_MESSAGE);
+                checkFalse(prepared.isTimestampSet(), NO_TIMESTAMPS_IN_UPDATES_MESSAGE);
+
+                preparedUpdates.add(prepared);
+            }
+
+            List<ConditionStatement> preparedConditions = new ArrayList<>(conditions.size());
+            for (ConditionStatement.Raw condition : conditions)
+                // TODO: Is this synthetic ks name dangerous?
+                preparedConditions.add(condition.prepare("[txn]", bindVariables));
+
+            return new TransactionStatement(preparedAssignments, returningSelect, returningReferences, preparedUpdates, preparedConditions, bindVariables);
+        }
+
+        private void checkAtMostOneRowSpecified(SelectStatement prepared, String failureMessage, Object messageArg)
+        {
+            int limit = prepared.getLimit(QueryOptions.DEFAULT);
+
+            if (limit == DataLimits.NO_LIMIT)
+                checkTrue(prepared.getRestrictions().hasAllPKColumnsRestrictedByEqualities(), failureMessage, messageArg);
+            else
+                checkTrue(limit == 1, failureMessage, messageArg);
```

Comment:
if I do `SELECT * FROM table where pk=? LIMIT 2` will this fail?  Looks like it will

Reply:
pushed a fix for this into the feedback branch

Reply:
going to resolve this; fixed in feedback branch and there is another comment that this is an issue (so don't need 2 comments)